blake2/src/impl_hacl.rs

```diff
@@ -0,0 +1,376 @@
+
+        let kk = index {
+            key_length,
+            digest_length,
+            last_node,
+        };
+
+        let params = blake2_params {
+            digest_length,
+            key_length,
+            fanout: 1,
+            depth: 1,
+            leaf_length: 0,
+            node_offset: 0,
+            node_depth: 0,
+            inner_length: 0,
+            salt: self.salt,
+            personal: self.personal,
+        };
+
+        let key = params_and_key {
+            fst: &[params],
+            snd: self.key,
+        };
+
+        Blake2b {
+            state: blake2b_malloc_raw(kk, key),
+        }
+    }
+}
+
+/// A builder for [`Blake2s`]. `KEY_LEN`` must be in the range `0..=64``, `OUT_LEN`` must be in the
+/// range `1..=64`.
+pub struct Blake2sBuilder<'a, const KEY_LEN: usize, const OUT_LEN: usize> {
+    key: &'a [u8; KEY_LEN],
+    personal: &'a [u8; BLAKE2S_PARAM_LEN],
+    salt: &'a [u8; BLAKE2S_PARAM_LEN],
+}
+
+impl<const OUT_LEN: usize> Blake2sBuilder<'_, 0, OUT_LEN> {
+    /// Creates the builder for an unkeyed hasher and returns an error if the digest length
+    /// `OUT_LEN` is not in the allowed range.
+    pub fn new_unkeyed() -> Result<Self, Error> {
+        if OUT_LEN < 1 || OUT_LEN > 64 {
+            return Err(Error::InvalidDigestLength);
+        }
+
+        Ok(Self {
+            key: &[],
+            personal: &[0; BLAKE2S_PARAM_LEN],
+            salt: &[0; BLAKE2S_PARAM_LEN],
+        })
+    }
+}
+impl<'a, const KEY_LEN: usize, const OUT_LEN: usize> Blake2sBuilder<'a, KEY_LEN, OUT_LEN> {
+    /// Creates the builder and returns an error if the lengths `KEY_LEN` or `OUT_LEN` are not in
+    /// the allowed range.
+    pub fn new() -> Result<Self, Error> {
+        if KEY_LEN > BLAKE2S_MAX_LEN {
+            return Err(Error::InvalidKeyLength);
+        }
+
+        if OUT_LEN < 1 || OUT_LEN > BLAKE2S_MAX_LEN {
+            return Err(Error::InvalidDigestLength);
+        }
+        Ok(Self {
+            key: &[0; KEY_LEN],
+            personal: &[0; BLAKE2S_PARAM_LEN],
+            salt: &[0; BLAKE2S_PARAM_LEN],
+        })
+    }
+
+    /// Sets the key to be used in the hasher.
+    pub fn with_key(self, key: &'a [u8; KEY_LEN]) -> Self {
+        Self { key, ..self }
+    }
+
+    /// Sets the personalization bytes to be used in the hasher.
+    pub fn with_personalization(self, personal: &'a [u8; BLAKE2S_PARAM_LEN]) -> Self {
+        Self { personal, ..self }
+    }
+
+    /// Sets the salt to be used in the hasher.
+    pub fn with_salt(self, salt: &'a [u8; BLAKE2S_PARAM_LEN]) -> Self {
+        Self { salt, ..self }
+    }
+
+    /// Constructs the [`Blake2s`] hasher.
+    pub fn build(self) -> Blake2s<KEY_LEN, OUT_LEN> {
+        // these are safe because they bot are at most 32
+        let key_length = KEY_LEN as u8;
+        let digest_length = OUT_LEN as u8;
+
+        // NOTE: I am not entirely sure that this is the correct value. From reading the spec I
```

Comment:
`false` is correct here. Aymeric wrote some notes when adding tests for `true` to hacl. https://github.com/hacl-star/hacl-star/pull/934. I don't think we need to expose the `true` version for now.